From e2d12daea395fcefe1f1b63af53517515598e428 Mon Sep 17 00:00:00 2001
From: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
Date: Thu, 8 Jun 2023 10:33:25 +0530
Subject: [PATCH] wifi: ath12k: fix the skb data when pkt offset present

when HTT metadata added in the skb data, the tx completion need
to remove this metadata and point the skb data to the actual payload
when its given to mac80211. Otherwise when this packet go through monitor
tx path then it face insufficient headroom and the packet get drops with
below warn_on().

Warn_on:
=======
WARNING: CPU: 3 PID: 2760 at net/mac80211/status.c:879 ieee80211_tx_monitor+0xf8/0x5d4 [mac80211]

Signed-off-by: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h  | 1 +
 drivers/net/wireless/ath/ath12k/dp_tx.c | 5 ++++-
 2 files changed, 5 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -118,6 +118,7 @@ struct ath12k_skb_cb {
 	struct ieee80211_vif *vif;
 	dma_addr_t paddr_ext_desc;
 	u8 link_id;
+	u16 pkt_offset;
 } __packed;
 
 struct ath12k_skb_rxcb {
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -556,6 +556,7 @@ int ath12k_dp_tx(struct ath12k *ar, stru
 	tx_desc->recycler_fast_xmit = 0;
 	ti.desc_id = tx_desc->desc_id;
 	ti.data_len = skb->len - ti.pkt_offset;
+	skb_cb->pkt_offset = ti.pkt_offset;
 	skb_cb->paddr = ti.paddr;
 	skb_cb->vif = ahvif->vif;
 
@@ -783,6 +784,9 @@ ath12k_dp_tx_htt_tx_complete_buf(struct
 			    skb_cb->link_id,
 			    ahvif->links_map);
 
+	if (skb_cb->pkt_offset)
+		skb_pull(msdu, skb_cb->pkt_offset);
+
 	memset(&info->status, 0, sizeof(info->status));
 
 	if (ts->acked) {
